crud/src/main/java/com/redhat/lightblue/assoc/RewriteQuery.java

```diff
@@ -0,0 +1,496 @@
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.assoc;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+
+import com.redhat.lightblue.metadata.CompositeMetadata;
+import com.redhat.lightblue.metadata.ResolvedReferenceField;
+import com.redhat.lightblue.metadata.FieldTreeNode;
+import com.redhat.lightblue.metadata.ArrayElement;
+
+import com.redhat.lightblue.query.*;
+
+import com.redhat.lightblue.util.Path;
+import com.redhat.lightblue.util.MutablePath;
+import com.redhat.lightblue.util.Error;
+import com.redhat.lightblue.util.CopyOnWriteIterator;
+
+/**
+ * Rewrite queries so that any clause referring to a field outside the
+ * current entity is removed from the query, and any clause comparing
+ * a field in this entity with a field in another entity is bound to a
+ * slot.
+ *
+ * It is not possible to rewrite a query clause referring to a field
+ * in another entity. For instance a query of the form x=1 would
+ * become meaningless, as 'x' has no value while evaluating this
+ * entity. So the clause is removed from the query, and replaced with
+ * a placeholder.
+ */
+public class RewriteQuery extends QueryIterator {
+
+    /**
+     * The root composite metadata
+     */
+    private final CompositeMetadata root;
+    
+    /**
+     * The entity for which the query is being rewritten
+     */
+    private final CompositeMetadata currentEntity;
+
+    /**
+     * Placeholder that always evaluates to true
+     */
+    public static final class TruePH extends QueryExpression {
+        @Override
+        public JsonNode toJson() {
+            return JsonNodeFactory.instance.booleanNode(true);
+        }
+    }
+
+    /**
+     * Placeholder that always evaluates to false
+     */
+    public static final class FalsePH extends QueryExpression {
+        @Override
+        public JsonNode toJson() {
+            return JsonNodeFactory.instance.booleanNode(false);
+        }
+    }
+
+    public static final class RewriteQueryResult {
+        public final QueryExpression query;
+        public final List<BoundObject> bindings;
+
+        public RewriteQueryResult(QueryExpression query,List<BoundObject> bindings) {
+            this.query=query;
+            this.bindings=bindings;
+        }
+    }
+
+    /**
+     * Construct a class to rewrite queries to retrieve an entity, potentially evaluated at another entity
+     *
+     * @param root The root entity
+     * @param correntEntity The query will be rewritten relative to this entity
```

Comment:
fixed.

On Tue, Mar 29, 2016 at 5:28 AM, Dennis Crissman <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/RewriteQuery.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57708050>
> :
>
> > +
> > +    public static final class RewriteQueryResult {
> > +        public final QueryExpression query;
> > +        public final List<BoundObject> bindings;
> > +
> > +        public RewriteQueryResult(QueryExpression query,List<BoundObject> bindings) {
> > +            this.query=query;
> > +            this.bindings=bindings;
> > +        }
> > +    }
> > +
> > +    /**
> > +     * Construct a class to rewrite queries to retrieve an entity, potentially evaluated at another entity
> > +     *
> > +     * @param root The root entity
> > +     * @param correntEntity The query will be rewritten relative to this entity
>
> nitpick: misspelled
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/4bcf56affb2ffd020fa4c380a7af065c40010abf#r57708050>
>